app/src/org/commcare/activities/DispatchActivity.java

```diff
@@ -333,23 +334,26 @@ private void handleUnvalidatedApp() {
     private void handleExternalLaunch() {
         //First off, make sure the incoming session is clear
         CommCareApplication.instance().getSession().proceedWithSavedSessionIfNeeded(() -> {
+                    Intent i = null;
                     if (getIntent().hasExtra(SESSION_REQUEST)) {
                         String sessionRequest = this.getIntent().getStringExtra(SESSION_REQUEST);
                         SessionStateDescriptor ssd = new SessionStateDescriptor();
                         ssd.fromBundle(sessionRequest);
                         CommCareApplication.instance().getCurrentSessionWrapper().loadFromStateDescription(ssd);
-                        Intent i = new Intent(this, StandardHomeActivity.class);
-                        i.putExtra(WAS_EXTERNAL, true);
-                        startActivityForResult(i, HOME_SCREEN);
+                        i = new Intent(this, StandardHomeActivity.class);
                     } else if (getIntent().hasExtra(SESSION_ENDPOINT_ID)) {
                         String sessionEndpointId = this.getIntent().getStringExtra(SESSION_ENDPOINT_ID);
                         Bundle args = this.getIntent().getBundleExtra(SESSION_ENDPOINT_ARGUMENTS_BUNDLE);
                         ArrayList<String> argsList = this.getIntent().getStringArrayListExtra(SESSION_ENDPOINT_ARGUMENTS_LIST);
-                        Intent i = new Intent(this, StandardHomeActivity.class);
-                        i.putExtra(WAS_EXTERNAL, true);
+                        i = new Intent(this, StandardHomeActivity.class);
                         i.putExtra(SESSION_ENDPOINT_ID, sessionEndpointId);
                         i.putExtra(SESSION_ENDPOINT_ARGUMENTS_BUNDLE, args);
                         i.putStringArrayListExtra(SESSION_ENDPOINT_ARGUMENTS_LIST, argsList);
+                    }
+                    if (i != null) {
+                        i.putExtra(WAS_EXTERNAL, true);
+                        i.putExtra(EXIT_AFTER_FORM_SUBMISSION,
+                                getIntent().getBooleanExtra(EXIT_AFTER_FORM_SUBMISSION, true));
```

Comment:
It's a small thing, but the 'true' default here is essentially set twice - once here and once in the base activity below. 

I think that could create some confusion later if someone though one of these two was the 'authoritative' default and didn't realize it was also getting defaulted upstream. 

Maybe it would be a little cleaner to check whether `EXIT_AFTER_FORM_SUBMISSION` exists in the calling intent and only set it here if so, to prevent the fallback and let the default in the `HomeScreenBaseActivity` be the only one?

Reply:
So like this - ?
```
if(getIntent().hasExtra(EXIT_AFTER_FORM_SUBMISSION)) {
                            i.putExtra(EXIT_AFTER_FORM_SUBMISSION,
                                    getIntent().getBooleanExtra(EXIT_AFTER_FORM_SUBMISSION, true));
                        }
```

even if we check for the extra, `getIntent().getBooleanExtra` requires a fallback. Would that be alright ?

Reply:
That does make sense. Implemented [here](https://github.com/dimagi/commcare-android/pull/2551)